File: src/algorithms/searching/binary/meta-binary-search/__tests__/meta-binary-search.test.ts
import { describe, it, expect } from "vitest";

import { metaBinarySearch } from "../sources/meta-binary-search.ts?fn";

describe("metaBinarySearch", () => {
  it("finds a value present in the array", () => {
    expect(metaBinarySearch([2, 5, 8, 12, 16, 23, 38, 56, 72, 91], 23)).toBe(5);
  });

  it("returns -1 when the value is not in the array", () => {
    expect(metaBinarySearch([2, 5, 8, 12, 16, 23, 38, 56, 72, 91], 50)).toBe(-1);
  });

  it("handles an empty array", () => {
    expect(metaBinarySearch([], 5)).toBe(-1);
  });

  it("handles a single element array when found", () => {
    expect(metaBinarySearch([42], 42)).toBe(0);
  });

  it("handles a single element array when not found", () => {
    expect(metaBinarySearch([42], 10)).toBe(-1);
  });

  it("finds the first element in the array", () => {
    expect(metaBinarySearch([2, 5, 8, 12, 16, 23, 38, 56, 72, 91], 2)).toBe(0);
  });

  it("finds the last element in the array", () => {
    expect(metaBinarySearch([2, 5, 8, 12, 16, 23, 38, 56, 72, 91], 91)).toBe(9);
  });

  it("finds the middle element in the array", () => {
    expect(metaBinarySearch([10, 20, 30, 40, 50], 30)).toBe(2);
  });

  it("returns -1 for a value smaller than all elements", () => {
    expect(metaBinarySearch([5, 10, 15, 20], 1)).toBe(-1);
  });

  it("returns -1 for a value larger than all elements", () => {
    expect(metaBinarySearch([5, 10, 15, 20], 100)).toBe(-1);
  });

  it("finds an element in a two element array", () => {
    expect(metaBinarySearch([3, 7], 7)).toBe(1);
  });

  it("handles an array of length that is a power of 2", () => {
    expect(metaBinarySearch([1, 3, 5, 7, 9, 11, 13, 15], 9)).toBe(4);
  });
});

File: src/algorithms/searching/binary/meta-binary-search/MetaBinarySearch_test.java
public class MetaBinarySearch_test {
    public static void main(String[] args) {
        int[] standardArray = {2, 5, 8, 12, 16, 23, 38, 56, 72, 91};

        assert MetaBinarySearch.metaBinarySearch(standardArray, 23) == 5 : "should find value present";
        assert MetaBinarySearch.metaBinarySearch(standardArray, 50) == -1 : "should return -1 when not found";
        assert MetaBinarySearch.metaBinarySearch(new int[]{}, 5) == -1 : "should handle empty array";
        assert MetaBinarySearch.metaBinarySearch(new int[]{42}, 42) == 0 : "should find single element";
        assert MetaBinarySearch.metaBinarySearch(new int[]{42}, 10) == -1 : "should return -1 for single element not found";
        assert MetaBinarySearch.metaBinarySearch(standardArray, 2) == 0 : "should find first element";
        assert MetaBinarySearch.metaBinarySearch(standardArray, 91) == 9 : "should find last element";
        assert MetaBinarySearch.metaBinarySearch(new int[]{10, 20, 30, 40, 50}, 30) == 2 : "should find middle element";
        assert MetaBinarySearch.metaBinarySearch(new int[]{5, 10, 15, 20}, 1) == -1 : "should return -1 for smaller than all";
        assert MetaBinarySearch.metaBinarySearch(new int[]{5, 10, 15, 20}, 100) == -1 : "should return -1 for larger than all";
        assert MetaBinarySearch.metaBinarySearch(new int[]{3, 7}, 7) == 1 : "should find element in two-element array";
        assert MetaBinarySearch.metaBinarySearch(new int[]{1, 3, 5, 7, 9, 11, 13, 15}, 9) == 4 : "should handle power-of-two length array";

        System.out.println("All tests passed!");
    }
}
